examples/index_md_into_pgvector.rs

```diff
@@ -0,0 +1,77 @@
+/**
+* This example demonstrates how to use the Pgvector integration with SwiftIDE
+*/
+use std::path::PathBuf;
+use swiftide::{
+    indexing::{
+        self,
+        loaders::FileLoader,
+        transformers::{
+            metadata_qa_text::NAME as METADATA_QA_TEXT_NAME, ChunkMarkdown, Embed, MetadataQAText,
+        },
+        EmbeddedField,
+    },
+    integrations::{self, pgvector::PgVector},
+};
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    tracing_subscriber::fmt::init();
+    tracing::info!("Starting PgVector indexing test");
+
+    // Get the manifest directory path
+    let manifest_dir = std::env::var("CARGO_MANIFEST_DIR").expect("CARGO_MANIFEST_DIR not set");
+
+    // Create a PathBuf to test dataset from the manifest directory
+    let test_dataset_path = PathBuf::from(manifest_dir).join("../README.md");
+
+    tracing::info!("Test Dataset path: {:?}", test_dataset_path);
+
+    let (_pgv_db_container, pgv_db_url, _temp_dir) = swiftide_test_utils::start_postgres().await;
+
+    tracing::info!("pgv_db_url :: {:#?}", pgv_db_url);
+
+    let llm_client = integrations::ollama::Ollama::default()
+        .with_default_prompt_model("llama3.2:latest")
+        .to_owned();
+
+    let fastembed =
+        integrations::fastembed::FastEmbed::try_default().expect("Could not create FastEmbed");
+
+    // Configure Pgvector with a default vector size, a single embedding
+    // and in addition to embedding the text metadata, also store it in a field
+    let pgv_storage = PgVector::builder()
+        .try_connect_to_pool(pgv_db_url, Some(10))
+        .await
+        .expect("Failed to connect to postgres server")
+        .vector_size(384)
+        .with_vector(EmbeddedField::Combined)
+        .with_metadata(METADATA_QA_TEXT_NAME)
+        .table_name("swiftide_pgvector_test".to_string())
+        .build()
+        .unwrap();
```

Comment:
Almost exactly right! I prefer it if builders do not do IO if they can avoid it, for multiple reasons. In this case, that also has the benefit of being able to connect lazilly and hiding the details of the connection pool. 

i.e. the builder api like:
```rust
 let pgv_storage = PgVector::builder()
        .database_url(pgv_db_url)
        .pool_size(10) // With a sane default if ommitted 
        .vector_size(384)
        .with_vector(EmbeddedField::Combined)
        .with_metadata(METADATA_QA_TEXT_NAME)
        .table_name("swiftide_pgvector_test".to_string())
        .build()
        .unwrap();
```

And then in `PgVector::setup` (which is only called once):

```rust
async fn setup(&self) -> Result<()> {
  self.try_connect_to_pool(self.database_url, self.pool_size).await?;
  ...
```

Reply:
@timonv, I'm looking for your input on a design choice here.

If we decide to handle the database connection pool setup within `fn setup(&self)` instead of `PgVectorBuilder`, we'll need to mutate `PgVector` within `fn setup()`. This change would mean updating the function signature in `trait Persist` to:

```rust
async fn setup(&mut self) -> Result<()>
```

For example:

```rust
async fn setup(&mut self) -> Result<()> {
  self.connection_pool = self.try_connect_to_pool(self.database_url, self.pool_size).await?;
  ...
}
```

This adjustment would introduce breaking changes across the stack, particularly impacting:

- `swiftide-indexing/src/persist/memory_storage.rs`
- `swiftide-integrations/src/lancedb/persist.rs`
- `swiftide-integrations/src/qdrant/persist.rs`
- `swiftide-integrations/src/redis/persist.rs`

Would you prefer moving the IO operations into `Persist::setup()` for these components? If so, we could handle this as a separate PR to streamline the updates.

Looking forward to your thoughts!

Reply:
Hmm good point. Since it runs only once and only during setup, perhaps inner mutability solves it nicer. I.e. with https://docs.rs/tokio/latest/tokio/sync/struct.OnceCell.html

Then:
```rust
async fn setup(&self) -> Result<()> {
  self.connection_pool.set(self.try_connect_to_pool().await?).await?;
  ...
}

// Where
struct PGVector {
  connection_pool: OnceCell<PGPool>
}

// And have a small wrapper to avoid the double error error handling like:
fn get_pool(&self) -> Result<&PGPool>
```

Reply:
@timonv, Thanks for the suggestion! It looks promising it should work well. I'll make the adjustments and update you shortly.

Reply:
Thank you for the valuagble suggestion, @timnov! I've completed the refactoring, and I believe the implementation is now cleaner and more efficient. 

Here’s a snippet of the updated code:

```rust
pub struct PgVector {
    ...

    /// Lazy-initialized database connection pool.
    #[builder(default = "Arc::new(OnceCell::new())")]
    connection_pool: Arc<OnceCell<PgPool>>,
}
```

Thanks again for the input!